### ISSUE TITLE
Problem with big number of fundMembers. 

### ISSUE THREAD

**Issue by astroseger** [ksridharbabuus/RFAI-Contracts#2]
In the current design there is the following problem. With sufficiently big number of fundMembers, we will no be able to call requestClaim, rejectRequest and closeRequest, because we will hit the  gas limit. It is actually a vector of attack.

The straightforward solution is following:
- We limit number of fundMembers
- We limit the minimal amount of funds (in order to prevent attack
with small stakes in order to hit our limit for fundMembers)

But I want to propose a slightly modified design in order to support unlimited number of fundMembers and submittedSols. It will also solve the problem with "div" in requestClaim (see item 12 in #1 ).

Our general target is to remove all for-loops with minimal change in functionality. To achieve this we should calculate funds which should be given to solutionSubmitter directly in “addFundsToRequest” and “vote” functions. It means that we should separately track amount which hasn't been yet distributed in "vote" and amounts which have been distributed in "vote".

1. We make two separate "vote" functions.
 - "vote_shortlist" -  for members to shortlist solutions.
 - "vote" - for everybody 

We change the semantic of ```vote``` function. In the vote function the voter should directly specify how his funds will be distributes among solutionSubmitters.
```
function vote(uint256 requestId, address [] solutionSubmitters,
uint256 [] funds) 
```
sum of ```funds``` should be equal to ```req.funds[msg.sender]```

User can change his mind and call vote again.

2. Unfortunately we will not be able to call settleFundsAndChangeStatus in rejectRequest and closeRequest. We should simply set expiration to 0. And users will need to call requestReClaim by themselves. I also propose to change name of requestReClaim to requestTakeBackMyStake.

**Comment by ksridharbabuus** [comment]
@astroseger My initial thought process is to calculate the claim during the voting itself and it became very complex to track the multiple times voting by the staker. Finally came up with the current logic.

In the current proposal the voter/staker will decide to distribute the funds which may not be a right option as we have to divide the funds equally to avoid any biased distribution.

I like the idea of limiting the minimum fund & the number of staking users. Will check as per the requirements and will update the logic accordingly.

**Comment by astroseger** [comment]
I think it is possible to make relatively simple design without "for" in requestClaim.

1. We have special struct for each stakeMember (instead of "funds")
```
struct stakeMemberStruct
{
      uint256 totalFunds;  //total amount (undesidedFunds + sum(fundsForVotedSubmitters))
      uint256  undesidedFunds; //funds which haven't been distributed among "voted" submitters. These funds will be evenly distributed across shortlisted solutions. 
      address[] votedSubmitters;   //list of submitters which was "voted" by this stakeMember 
      uint256[]  fundsForVotedSubmitters; //list of corresponding funds.
}
``` 

2. We have corresponding "global" values inside "Request", but globalFundsForVotedSubmitters will be a mapping.
```
uint256 globalTotalFunds
uint256 globalundesidedFunds
mapping(address => unt256) globalfundsForVotedSubmitters
```

3. In "addFundsToRequest", we simply add funds to undesidedFunds (and we update globalTotalFunds!)

4. We make the following vote function: 
```
//sum(fundsForVotedSubmitters) should be equal to yours totalFunds
function vote(uint256 requestId, address [] votedSubmitters, uint256 [] fundsForVotedSubmitters);
```
So we simply copy votedSubmitters, fundsForVotedSubmitters and set undesidedFunds to 0. We also update "global" variables accordingly: subtract old fundsForVotedSubmitters and add new fundsForVotedSubmitters. 

5. We have separate list for shortlisted solutions which foundation member can update via special vote function (vote_shortlisting)

**Comment by ksridharbabuus** [comment]
@astroseger unfortumately we cannot avoid the loops.
For example I need to check 
All the votedSubmitters passed in the array are valid submitter.
Sum of the funds distributed should be equal to the staker funds

I tried the current implementation with less number of structs to reduce the storage. Will try implementing a different approach to avoid the loops for the funds claim using different structure which I have I tried earlier.

**Comment by astroseger** [comment]
Yes! We cannot remove loops completely, but we can remove loops over all stakeMembers. So we could support infinite number of stakeMembers. Actually I'm not 100% sure that we need it (but we might need it)... 

The current solution is totally fine as it is, but with limited number of stakeMembers... 
The problem with any solution with unlimited number of stakeMembers is following: we will not be able to run settleFundsAndChangeStatus in rejectRequest and closeRequest, so users will need to call requestTakeBackMyStake by themselves.

**Comment by ksridharbabuus** [comment]
@astroseger Tested today upto 10 StakeMembers and it is working with no issues. Yet to check with Raam on the number and will limit based on the requirements.

**Comment by astroseger** [comment]
10 is too small. I though that initial plan was to make something similar to crowdfunding.  So we need much more StakeMembers. We might need unlimited number of StakeMembers...  I think we need to discuss with Raam and Cassio.

**Comment by ksridharbabuus** [comment]
Sorry typo mistake in the number. Tested with 100 stake members. And seems
to be no issue. Will continue testing with higher numbers today as well.
Will discuss with raam and will add the limits.

**Comment by ksridharbabuus** [comment]
@astroseger Added the limits to minStake and MaxStakers in the contract.

**Comment by astroseger** [comment]
What will be the limit for MaxStakers on main net? Do Raam and Cassio agree on solution with limited number of Stakers?
Why you didn't like solution with unlimited number of Stakers?  (Probably we should say that solution with unlimited number of Stakers exists, but it will require some work to implement it... )

**Comment by ksridharbabuus** [comment]
Tried the options and it getting more complex from execution.
Also we are related on the user to provide distribution which will not be
best option as the split need to be by  voting and automatically
distributes based on votes.
Discussed with Raam and will be sharing the same with Casio...

**Comment by astroseger** [comment]
@ksridharbabuus 
In the current design user also "provide distribution" by voting several times. Actually the solution which I've proposed more general, because you can simulate current voting. You can get the current "distribution", modify it according to your new vote and submit it again. :)

Yes. now I see that solution with "providing distribution" is better because:
1. We can have infinite number of stakers
2. We can fully simulate the current design on the client side (with exception of settleFundsAndChangeStatus, so stakers will need to claim funds by themselves after rejectRequest and closeRequest)
